draft-ietf-tls-esni.md

```diff
@@ -324,10 +324,10 @@ cipher_suite
 provided in the corresponding `ECHConfig.cipher_suites` list.
 
 config_digest
-: A cryptographic hash of the `ECHConfig` structure from which the ECH key was
-obtained, i.e., from the first byte of "version" to the end of the structure.
-This hash is computed using the hash function associated with `cipher_suite`,
-i.e., the corresponding HPKE KDF algorithm hash.
+: Equal to `Extract(Expand(salt, config), "ech_config_digest", Nh)`, where
+`salt` is a string of `Nh` bytes set to `0`, `config` is the `ECHConfig`
+structure, and `Extract`, `Expand`, and `Nh` are as defined by the KDF API in
+{{!I-D.irtf-cfrg-hpke}}.
```

Comment:
Is the empty string <=> all zero thing a deliberate choice or just an odd quirk of HKDF (via HMAC)?  I agree that in this case the difference isn't important and an empty string is fine, but do we need to be careful about this at all?

Reply:
For HKDF it's the same (if salt's omitted, HKDF just fills in a zero-length string of Nh bytes), but for other KDFs it might not be. So we decided to just stick with empty strings, and I think we ought to do the same here.

Reply:
[You have to be careful with the salt length for HKDF](https://hal.inria.fr/hal-02100345). As I recall, a safe choice is the output length of the underlying hash function. (This is `Nh` in the HPKE draft.) This is also what's done in RFC 8446.

Reply:
Quoting from the HKDF RFC:

> salt: if not provided, it is set to a string of HashLen zeros.

What I'm proposing is that we omit this, since we don't want HKDF quirks impacting future KDFs. (This is a deliberate design choice we made for HPKE.)

Reply:
Is "" the same as Nh-length byte stering? I read "" as the 0-length string.

Reply:
It's an empty string, i.e., equivalent to not providing the input.

Reply:
What does it look like, syntactically, if the salt isn't provided? Were I reading this, I would think that the salt is the empty string.

Possible variants:
- `Expand(salt, ikm)` where `salt` is ...
- `Expand("", ikm)`
- `Expand(0, ikm)`

Reply:
Chatted offline -- `Extract("", ikm)` is the way to go!

Reply:
Will do with this in the morning!

Reply:
We'll go with "" == "\x00" * Nh, here and below. I'll add a paranthetical to clarify this the first time we use this.

Reply:
Done.